File: VB/Dennis.Linq/LinqCollectionSource.vb
' Developer Express Code Central Example:
' How to populate the list view with data from a LINQ query
' 
' See the http://www.devexpress.com/scid=K18107 KB Article for more information.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E859


Imports Microsoft.VisualBasic
Imports System
Imports System.Linq
Imports DevExpress.Xpo
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Data.Filtering

Namespace Dennis.Linq
	Public Class LinqCollectionSource
		Inherits CollectionSourceBase
		Public Const DefaultSuffix As String = "_Linq"
		Private collectionCore As IBindingList
		Private objectTypeInfoCore As ITypeInfo
		Public Function ConvertQueryToCollection(ByVal sourceQuery As IQueryable) As IList
			collectionCore = New BindingList(Of Object)()
			For Each item In sourceQuery
				collectionCore.Add(item)
			Next item
			Return collectionCore
		End Function
		Private queryCore As IQueryable = Nothing
		Protected Sub New(ByVal objectSpace As IObjectSpace, ByVal mode As CollectionSourceMode)
			MyBase.New(objectSpace, mode)
		End Sub
		Protected Sub New(ByVal objectSpace As IObjectSpace)
			MyBase.New(objectSpace)
		End Sub
		Public Sub New(ByVal objectSpace As IObjectSpace, ByVal objectType As Type, ByVal query As IQueryable)
			MyBase.New(objectSpace)
			objectTypeInfoCore = XafTypesInfo.Instance.FindTypeInfo(objectType)
			queryCore = query
		End Sub
		Public Property Query() As IQueryable
			Get
				Return queryCore
			End Get
			Set(ByVal value As IQueryable)
				queryCore = value
			End Set
		End Property
		Protected Overrides Function CreateCollection() As Object
			CType(Query, XPQueryBase).Session = (CType(ObjectSpace, ObjectSpace)).Session
			Return ConvertQueryToCollection(Query)
		End Function
		Public Overrides Function IsObjectFitForCollection(ByVal obj As Object) As Boolean?
			Return collectionCore.Contains(obj)
		End Function
		Protected Overrides Sub ApplyCriteriaCore(ByVal criteria As CriteriaOperator)
		End Sub
		Public Overrides ReadOnly Property ObjectTypeInfo() As ITypeInfo
			Get
				Return objectTypeInfoCore
			End Get
		End Property
	End Class
End Namespace

File: CS/Dennis.Linq/LinqCollectionSource.cs
// Developer Express Code Central Example:
// How to populate the list view with data from a LINQ query
// 
// See the http://www.devexpress.com/scid=K18107 KB Article for more information.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E859

using System;
using System.Linq;
using DevExpress.Xpo;
using System.Collections;
using System.ComponentModel;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Xpo;

namespace Dennis.Linq {
    public class LinqCollectionSource : CollectionSourceBase {
        public const string DefaultSuffix = "_Linq";
        private IBindingList collectionCore;
        private ITypeInfo objectTypeInfoCore;
        public IList ConvertQueryToCollection(IQueryable sourceQuery) {
            collectionCore = new BindingList<object>();
            foreach (var item in sourceQuery) { collectionCore.Add(item); }
            return collectionCore;
        }
        private IQueryable queryCore = null;
        protected LinqCollectionSource(IObjectSpace objectSpace, CollectionSourceMode mode)
            : base(objectSpace, mode) {
        }
        protected LinqCollectionSource(IObjectSpace objectSpace)
            : base(objectSpace) {
        }
        public LinqCollectionSource(IObjectSpace objectSpace, Type objectType, IQueryable query)
            : base(objectSpace) {
            objectTypeInfoCore = XafTypesInfo.Instance.FindTypeInfo(objectType);
            queryCore = query;
        }
        public IQueryable Query {
            get { return queryCore; }
            set { queryCore = value; }
        }
        protected override object CreateCollection() {
            ((XPQueryBase)Query).Session = ((XPObjectSpace)ObjectSpace).Session;
            return ConvertQueryToCollection(Query);
        }
        public override bool? IsObjectFitForCollection(object obj) {
            return collectionCore.Contains(obj);
        }
        protected override void ApplyCriteriaCore(CriteriaOperator criteria) { }
        public override ITypeInfo ObjectTypeInfo {
            get { return objectTypeInfoCore; }
        }
    }
}